### ISSUE TITLE
[问题/Issue] 章节2.3：搭建Transformer的代码里有错误

### ISSUE THREAD

**Issue by ljhzttljh** [datawhalechina/happy-llm#191]
### 1. 遇到问题的章节 / Affected Chapter

Chapter 2.3

### 2. 具体问题描述 / Problem Description

搭建Transformer的代码里，的forward函数里面的Encoder和Decoder的输入是同一个x

### 3. 问题重现材料 / Reproduction Materials

def forward(self, idx, targets=None):
        # 输入为 idx，维度为 (batch size, sequence length, 1)；targets 为目标序列，用于计算 loss
        device = idx.device
        b, t = idx.size()
        assert t <= self.args.block_size, f"不能计算该序列，该序列长度为 {t}, 最大序列长度只有 {self.args.block_size}"

        # 通过 self.transformer
        # 首先将输入 idx 通过 Embedding 层，得到维度为 (batch size, sequence length, n_embd)
        print("idx",idx.size())
        # 通过 Embedding 层
        tok_emb = self.transformer.wte(idx)
        print("tok_emb",tok_emb.size())
        # 然后通过位置编码
        pos_emb = self.transformer.wpe(tok_emb) 
        # 再进行 Dropout
        **x = self.transformer.drop(pos_emb)**
        # 然后通过 Encoder
        print("x after wpe:",x.size())
        **enc_out = self.transformer.encoder(x)**
        print("enc_out:",enc_out.size())
        # 再通过 Decoder
        **x = self.transformer.decoder(x, enc_out)**
        print("x after decoder:",x.size())

        if targets is not None:
            # 训练阶段，如果我们给了 targets，就计算 loss
            # 先通过最后的 Linear 层，得到维度为 (batch size, sequence length, vocab size)
            logits = self.lm_head(x)
            # 再跟 targets 计算交叉熵
            loss = F.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1), ignore_index=-1)
        else:
            # 推理阶段，我们只需要 logits，loss 为 None
            # 取 -1 是只取序列中的最后一个作为输出
            logits = self.lm_head(x[:, [-1], :]) # note: using list [-1] to preserve the time dim
            loss = None

        return logits, loss

### 确认事项 / Verification

- [x] 此问题未在过往Issue中被报告过 / This issue hasn't been reported before

**Comment by ksk2023** [comment]
验证了这个问题，确实存在。

**确认**：原 `forward` 中 `enc_out = encoder(x)` 后又 `decoder(x, enc_out)` —— Decoder 的输入与 Encoder 共用同一个 `x`（含位置编码后的源序列表示），语义错误。标准 Encoder-Decoder 中 Decoder 应对目标序列单独做 Embedding + 位置编码，再以 `enc_out` 作为交叉注意力的 K/V。

注意到 PR #210 已包含此修复（`forward(idx, ...)` 改为 `forward(src, tgt, ...)`，src/tgt 分别编码），同时一并修复了 #199 之外的维度问题与公式索引，可直接关联本 issue，期待维护者 review 合并～
